### PR TITLE
Support for casting with DECIMAL types

### PR DESCRIPTION
**Pull request by agavra**
### Description 
This commit adds support for comparing decimals.

### Testing done 
- Unit testing
- QTT tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by agavra** [comment]
> The cast stuff LGTM. Is there a separate PR for casting from DECIMAL to other numeric types?

Yup! That'll be next, perhaps alongside math with other numeric types.

File: ksql-common/src/main/java/io/confluent/ksql/util/DecimalUtil.java [hunk 1]
```diff
@@ -142,6 +138,47 @@ public static BigDecimal ensureFit(final BigDecimal value, final Schema schema)
...
+    validateParameters(precision, scale);
+    final BigDecimal decimal = new BigDecimal(value);
+    ensureMax(decimal, precision, scale);
+    return decimal.setScale(scale, RoundingMode.HALF_UP);
+  }
+
+  private static void ensureMax(final BigDecimal value, final int precision, final int scale) {
+    final int digits = precision - scale;
+    final BigDecimal maxValue = new BigDecimal(Math.pow(10, digits));
+    if (maxValue.compareTo(value) < 1) {
```
Comment (big-andy-coates) [inline ksql-common/src/main/java/io/confluent/ksql/util/DecimalUtil.java]: as discussed - we need some test cases for negative numbers ;)

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java [hunk 2]
```diff
@@ -418,12 +420,15 @@ private String visitBooleanComparisonExpression(final ComparisonExpression.Type
       return new Pair<>(expr, Schema.OPTIONAL_BOOLEAN_SCHEMA);
     }
 
+    // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: Noooooo........ this rule is generally a clear indication the logic of the function is too complex to easily grok.  Don't disable the rule, instead refactor, e.g. move your decimal handling into `visitDecimalCast`, or better still replace the switch with code that looks up the handler from a static map.

`SqlToJavaVisitor` is a bit of a god class - we need to think about how to break it up.  However in the mean time, maybe handle the cast functionality in a utility class, and just call that from here.  This encourages people to break up the functionality into small targeted methods that are easy to understand... and improves test coverage!

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java [hunk 3]
```diff
@@ -799,6 +813,30 @@ private String getCastString(
     }
   }
 
+  private String getDecimalCastString(
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: nit: static?

File: ksql-engine/src/main/java/io/confluent/ksql/util/ExpressionTypeManager.java [hunk 4]
```diff
@@ -110,7 +111,7 @@ protected Expression visitCast(
       final ExpressionTypeContext expressionTypeContext
   ) {
     final Type sqlType = node.getType();
-    if (!(sqlType instanceof PrimitiveType)) {
+    if (!(sqlType instanceof PrimitiveType) && !(sqlType instanceof Decimal)) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/ExpressionTypeManager.java]: Is it worth having a marker interface? e.g. have `PrimitveType` and `Decimal` implement `Castable` interface?

Also, add decimals to the message in the exception

File: ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java [hunk 5]
```diff
@@ -496,6 +498,21 @@ public void shouldGenerateCorrectCodeForDecimalUnaryPlus() {
     assertThat(java, is("(TEST1_COL8.plus(new MathContext(2, RoundingMode.UNNECESSARY)))"));
   }
 
+  @Test
+  public void shouldGenerateCorrectCodeForDecimalCast() {
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java]: how about a negative tests that test the two if statements you've added/changed in the class?

Also, is it worth having a test to ensure CAST  from decimal to a decimal with the same scale and precision is a no op?

(Easier if you pull out the logic into a util and add the tests there...)

File: ksql-functional-tests/src/test/resources/query-validation-tests/cast.json [hunk 6]
```diff
@@ -61,6 +61,88 @@
...
+        {"topic":  "OUT", "value": {"VAL": "0.1"}},
+        {"topic":  "OUT", "value": {"VAL": "1.1"}},
+        {"topic":  "OUT", "value": {"VAL": "0.0"}},
+        {"topic":  "OUT", "value": {"VAL": "1.0"}},
+        {"topic":  "OUT", "value": {"VAL": "0.1"}},
+        {"topic":  "OUT", "value": {"VAL": "0.0"}}
+      ]
+    },
+    {
+      "name": "string to decimal",
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/cast.json]: Consider adding negative tests for casting numbers too large, (+'ve and -'ve), as these test what error the user sees. Same for cast to invalid type.

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java (hunk 2 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: > e.g. move your decimal handling into visitDecimalCast

did this I guess after you had written your comment :) but it's still not good enough for checkstyle

> or better still replace the switch with code that looks up the handler from a static map.

okay you caught me being lazy

File: ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java (hunk 5 above)
Comment (agavra) [reply ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java]: the last bit can be tested in `DecimalUtilTest` - I have added a test for that

File: ksql-functional-tests/src/test/resources/query-validation-tests/cast.json (hunk 6 above)
Comment (agavra) [reply ksql-functional-tests/src/test/resources/query-validation-tests/cast.json]: unfortunately errors in casting end up just being swallowed and `null` values are returned, so it's not especially useful as a QTT. improving the user experience around bad values is out of scope unfortunately

File: ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java (hunk 5 above)
Comment (agavra) [reply ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java]: The former is impossible - we will never have a SqlType that resolves to `DECIMAL` and whose type is not decimal (so impossible I can't test it without mocking a bizzaro type). I think that's a little bit of test bloat.

File: ksql-functional-tests/src/test/resources/query-validation-tests/cast.json (hunk 6 above)
Comment (big-andy-coates) [reply ksql-functional-tests/src/test/resources/query-validation-tests/cast.json]: Good point, and agreed!

File: ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java (hunk 5 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java]: For the former... if its impossible, why is there an if block for it? :D

Comment (agavra) [reply ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java]: so IntelliJ doesn't complain about unchecked casts :P

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: Yup, I did. =)

I reckon the best approach is to pull the full function out into a util class and then refactor it and test it.  but that is more work...

Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: Testing it in isolation is a little weird - I just refactored it into an internal static class and used a handler map

File: ksql-functional-tests/src/test/resources/query-validation-tests/cast.json (hunk 6 above)
Comment (agavra) [reply ksql-functional-tests/src/test/resources/query-validation-tests/cast.json]: I might actually be wrong about that - I just noticed that `cast.json` already has some exceptional cases. I guess maybe only `KsqlFunctionalException` gets raised? Now the question is whether in the case of a bad cast it's better to return `null` as with most KSQL errors or to fail hard...

**Pull request merged**